Invalid TypeScript emitted when rule has arguments

From this grammar:
https://github.com/prestodb/presto/blob/master/presto-parser/src/main/antlr4/com/facebook/presto/sql/parser/SqlBase.g4


Result:
```
$ ./antlr4ts src/simulator/sql/SqlBase.g4 
Generating file '/Users/carl/antlr/src/simulator/sql/SqlBaseLexer.ts' for grammar 'src/simulator/sql/SqlBase.g4'
Generating file '/Users/carl/antlr/src/simulator/sql/SqlBaseLexer.tokens' for grammar 'src/simulator/sql/SqlBase.g4'
Generating file '/Users/carl/antlr/src/simulator/sql/SqlBaseParser.ts' for grammar 'src/simulator/sql/SqlBase.g4'
Generating file '/Users/carl/antlr/src/simulator/sql/SqlBaseListener.ts' for grammar 'src/simulator/sql/SqlBase.g4'
Generating file '/Users/carl/antlr/src/simulator/sql/SqlBase.tokens' for grammar 'src/simulator/sql/SqlBase.g4'
$ tsc
src/simulator/sql/SqlBaseParser.ts(10711,2): error TS2394: Overload signature is not compatible with function implementation.
src/simulator/sql/SqlBaseParser.ts(10714,66): error TS1016: A required parameter cannot follow an optional parameter.
```

The problematic generated code:
```ts
export class PredicateContext extends ParserRuleContext {
	public value: ParserRuleContext;
	constructor();
	constructor(parent: ParserRuleContext, invokingState: number);
	constructor(parent: ParserRuleContext, invokingState: number, value: ParserRuleContext);
	constructor(parent?: ParserRuleContext, invokingState?: number, value: ParserRuleContext) {
		if (parent !== undefined && invokingState !== undefined) {
			super(parent, invokingState);
		} else {
			super();
		}
		this.value = value;
	}
```

fyi, the latest from 0.4.0-alpha.4...  `tsc` (2.1.5) fails on the generated code (from Presto's SqlBase.g4) with the following:

```
src/SqlBaseParser.ts(6456,3): error TS2345: Argument of type 'TypedPropertyDescriptor<() => WindowFrameContext>' is not assignable to parameter of type 'TypedPropertyDescriptor<(...args: any[]) => ParserRuleContext>'.
  Type '() => WindowFrameContext' is not assignable to type '(...args: any[]) => ParserRuleContext'.
    Type 'WindowFrameContext' is not assignable to type 'ParserRuleContext'.
      Types of property '_start' are incompatible.
        Type 'FrameBoundContext' is not assignable to type 'Token'.
          Property 'type' is missing in type 'FrameBoundContext'.
src/SqlBaseParser.ts(6459,18): error TS2345: Argument of type 'WindowFrameContext' is not assignable to parameter of type 'ParserRuleContext'.
src/SqlBaseParser.ts(6465,24): error TS2345: Argument of type 'WindowFrameContext' is not assignable to parameter of type 'ParserRuleContext'.
src/SqlBaseParser.ts(6475,24): error TS2345: Argument of type 'WindowFrameContext' is not assignable to parameter of type 'ParserRuleContext'.
src/SqlBaseParser.ts(6485,24): error TS2345: Argument of type 'WindowFrameContext' is not assignable to parameter of type 'ParserRuleContext'.
src/SqlBaseParser.ts(6501,24): error TS2345: Argument of type 'WindowFrameContext' is not assignable to parameter of type 'ParserRuleContext'.
src/SqlBaseParser.ts(10936,2): error TS2394: Overload signature is not compatible with function implementation.
src/SqlBaseParser.ts(10939,66): error TS1016: A required parameter cannot follow an optional parameter.
src/SqlBaseParser.ts(12081,36): error TS2345: Argument of type 'typeof WindowFrameContext' is not assignable to parameter of type 'new (...args: any[]) => ParserRuleContext'.
  Type 'WindowFrameContext' is not assignable to type 'ParserRuleContext'.
src/SqlBaseParser.ts(12109,14): error TS2415: Class 'WindowFrameContext' incorrectly extends base class 'ParserRuleContext'.
  Types of property '_start' are incompatible.
    Type 'FrameBoundContext' is not assignable to type 'Token'.
```

--- a/tool/resources/org/antlr/v4/tool/templates/codegen/TypeScript/TypeScript.stg
+++ b/tool/resources/org/antlr/v4/tool/templates/codegen/TypeScript/TypeScript.stg
@@ -760,22 +760,8 @@ StructDecl(struct,ctorAttrs,attrs,getters,dispatchMethods,interfaces,extensionMe
 export class <struct.name> extends <if(contextSuperClass)><contextSuperClass><else>ParserRuleContext<endif><if(interfaces)> implements <interfaces; separator=", "><endif> {
 	<attrs:{a | public <a>;}; separator="\n">
 	<getters:{g | <g>}; separator="\n">
-<if(struct.provideCopyFrom)>
-	constructor();
-<endif>
-	constructor(parent: ParserRuleContext, invokingState: number<ctorAttrs:{a | , <a>}>);
-<if(struct.provideCopyFrom)>
-	constructor(parent?: ParserRuleContext, invokingState?: number<ctorAttrs:{a | , <a>}>) {
-		if (parent !== undefined && invokingState !== undefined) {
-			super(parent, invokingState);
-		} else {
-			super();
-		}
-<else>
-	constructor(parent: ParserRuleContext, invokingState: number<ctorAttrs:{a | , <a>}>) {
+	constructor(parent: ParserRuleContext | undefined, invokingState: number<ctorAttrs:{a | , <a>}>) {
 		super(parent, invokingState);
-<endif>
-
 		<struct.ctorAttrs:{a | this.<a.name> = <a.name>;}; separator="\n">
 	}
 	@Override public get ruleIndex(): number { return <parser.name>.RULE_<struct.derivedFromName>; }
@@ -794,7 +780,10 @@ AltLabelStructDecl(struct,attrs,getters,dispatchMethods) ::= <<
 export class <struct.name> extends <currentRule.currentRule.name; format="cap">Context {
 	<attrs:{a | public <a>;}; separator="\n">
 	<getters:{g | <g>}; separator="\n">
-	constructor(ctx: <currentRule.currentRule.name; format="cap">Context) { super(); this.copyFrom(ctx); }
+	constructor(ctx: <currentRule.currentRule.name; format="cap">Context) {
+		super(ctx.parent, ctx.invokingState<currentRule.ruleCtx.ctorAttrs:{a | , ctx.<a.d.name>}>);
+		this.copyFrom(ctx);
+	}
 	<dispatchMethods; separator="\n">
 }
 >>

--- a/tool/resources/org/antlr/v4/tool/templates/codegen/TypeScript/TypeScript.stg
+++ b/tool/resources/org/antlr/v4/tool/templates/codegen/TypeScript/TypeScript.stg
@@ -763,9 +763,6 @@ export class <struct.name> extends <if(contextSuperClass)><contextSuperClass><el
 <if(struct.provideCopyFrom)>
 	constructor();
 <endif>
-<if(ctorAttrs)>
-	constructor(parent: ParserRuleContext, invokingState: number);
-<endif>
 	constructor(parent: ParserRuleContext, invokingState: number<ctorAttrs:{a | , <a>}>);
 <if(struct.provideCopyFrom)>
 	constructor(parent?: ParserRuleContext, invokingState?: number<ctorAttrs:{a | , <a>}>) {

--- a/tool/test/org/antlr/v4/test/runtime/typescript/TypeScript.test.stg
+++ b/tool/test/org/antlr/v4/test/runtime/typescript/TypeScript.test.stg
@@ -205,7 +205,7 @@ Column() ::= "this.charPositionInLine"
 
 Text() ::= "this.text"
 
-ValEquals(a,b) ::= <%<a>==<b>%>
+ValEquals(a,b) ::= <%<a> === <b>%>
 
 TextEquals(a) ::= <%this.text === "<a>"%>
 
@@ -628,17 +628,6 @@ IgnoredTests ::= [
 	// Skipped due to limitations in the testing framework
 
 	// Skipped due to limitations in the TypeScript target
-	"CompositeParsers.DelegatorInvokesDelegateRuleWithArgs": "Arguments are not supported with labeled outer alternatives.",
-
-	"SemPredEvalParser.AtomWithClosureInTranslatedLRRule": "Arguments are not supported with labeled outer alternatives.",
-	"SemPredEvalParser.DepedentPredsInGlobalFOLLOW": "Arguments are not supported with labeled outer alternatives.",
-	"SemPredEvalParser.DependentPredNotInOuterCtxShouldBeIgnored": "Arguments are not supported with labeled outer alternatives.",
-	"SemPredEvalParser.PredicateDependentOnArg": "Arguments are not supported with labeled outer alternatives.",
-	"SemPredEvalParser.PredicateDependentOnArg2": "Arguments are not supported with labeled outer alternatives.",
-
-	"FullContextParsing.ExprAmbiguity_1": "Arguments are not supported with labeled outer alternatives.",
-	"FullContextParsing.ExprAmbiguity_2": "Arguments are not supported with labeled outer alternatives.",
-
 	"Performance.DropLoopEntryBranchInLRRule_4": "Need a timeout of 15000ms for this to pass.",
 
 	default: false

--- a/tool/test/org/antlr/v4/test/runtime/typescript/TypeScript.test.stg
+++ b/tool/test/org/antlr/v4/test/runtime/typescript/TypeScript.test.stg
@@ -374,7 +374,7 @@ import { ParserRuleContext as PRC } from "antlr4ts/ParserRuleContext";
 
 export class MyRuleNode extends PRC {
 	private altNum: number;
-	constructor(parent: PRC, invokingStateNumber: number) {
+	constructor(parent: PRC | undefined, invokingStateNumber: number) {
 		super(parent, invokingStateNumber);
 	}
 

